spec/sneakers/integration_spec.rb

```diff
@@ -100,8 +85,352 @@ def start_worker(w)
+      cleanup_redis(redis)
+      prepare_sneakers(exchange: exchange_name)
+
+      worker.from_queue(queue_name, worker_opts)
+      @worker_pid = start_worker(worker)
+    end
+
+    after :each do
+      Process.kill('TERM', @worker_pid)
+
+      cleanup_rabbitmq(rabbitmq_client)
+      cleanup_redis(redis)
+    end
+
+    describe 'with defaults for maxretry handler' do
+      let(:worker) { MaxretryWorker }
+      let(:worker_opts) do
+        {
+          handler:         Sneakers::Handlers::Maxretry,
+          retry_max_times: 2,
+          retry_timeout:   100,
+          arguments: {
+            'x-dead-letter-exchange' => "#{queue_name}-retry"
+          }
+        }
+      end
+
+      it 'creates the required exchanges' do
+        expected_exchange_names = [
+          exchange_name,
+          "#{queue_name}-error",
+          "#{queue_name}-retry",
+          "#{queue_name}-retry-requeue"
+        ]
+        exchange_names = rabbitmq_client.list_exchanges.map(&:name)
+
+        expected_exchange_names.each do |expected_exchange_name|
+          assert_includes(exchange_names, expected_exchange_name)
+        end
+      end
+
+      it 'creates the required queues' do
+        expected_queue_names = [
+          queue_name,
+          "#{queue_name}-error",
+          "#{queue_name}-retry"
+        ]
+        queue_names = rabbitmq_client.list_queues.map(&:name)
+
+        expected_queue_names.each do |expected_queue_name|
+          assert_includes(queue_names, expected_queue_name)
+        end
+      end
+
+      describe 'when worker allways fails' do
+        let(:worker) { FailingMaxretryWorker }
+
+        before do
+          Sneakers::Publisher.new(
+            exchange: exchange_name
+          ).publish(
+            'foo',
+            routing_key: queue_name
+          )
+        end
+
+        it 'has a message in the error queue' do
+          # wait for failing message
+          sleep REQUEUING_WAIT_TIME
+
+          message = get_message_from_queue("#{queue_name}-error")
+
+          refute_nil(message.first)
+        end
+
+        it 'has been retried twice' do
+          # wait for failing message
+          sleep REQUEUING_WAIT_TIME
+
+          message = get_message_from_queue("#{queue_name}-error")
+
+          # Hmmmmm... should be 2
+          assert_equal(JSON.load(message[2])['num_attempts'], 3)
+        end
+      end
+
+      describe 'when worker fails once' do
+        let(:worker) { RetryOnceWorker }
+
+        before do
+          Sneakers::Publisher.new(
+            exchange: exchange_name
+          ).publish(
+            'foo',
+            routing_key: queue_name
+          )
+        end
+
+        it 'has no message in the error queue' do
+          # wait for failing message
+          sleep REQUEUING_WAIT_TIME
+
+          message = get_message_from_queue("#{queue_name}-error")
+
+          assert_nil(message.first)
+        end
+
+        it 'consumes the requeued message' do
```

Comment:
Ok, now I got your point. I never thought about a scenario where routing keys are important for the consumer. I see routing keys as information for rabbitmq. The worker needs to know the routing key, only because it creates the binding between exchange and queue. It consumes every message in the queue no matter how it gets there.
The retried message does not have the same routing key as the original message. There is an additional binding which routes the retry messages to the original queue.
I added some integration tests to show how the maxretry handler would work in our existing environment. It creates only one additional exchange. The error exchange. This is why we need different routing keys.
If it is important to keep the original routing key, I think I should add a new
handler instead of changing this one.